File: src/voz.py
"""
Registro canonico de vozes.

A geracao visual ja usa locks fortes de personagem. Este modulo aplica a
mesma ideia ao audio: cada personagem recebe uma identidade vocal fixa e as
cenas apenas copiam esse contrato.
"""

import re

from dialogo import lock_idioma_texto, normalizar_idioma


def chave_speaker(nome):
    return re.sub(r"[^A-Z0-9]+", "_", str(nome or "").upper()).strip("_")


def _genero_canonico(gender):
    texto = str(gender or "").lower()
    if any(x in texto for x in ("female", "femin", "mulher", "girl", "woman")):
        return "Strictly Female"
    if any(x in texto for x in ("male", "masc", "homem", "boy", "man")):
        return "Strictly Male"
    return "Androgynous / Character Voice"


def _sintese_forcada(gender, idioma):
    genero = _genero_canonico(gender)
    codigo = normalizar_idioma(idioma).upper()
    if "Female" in genero:
        return f"Female_Adult_Soprano_{codigo}"
    if "Male" in genero:
        return f"Male_Adult_Baritone_{codigo}"
    return f"Adult_Neutral_Character_{codigo}"


def _peso_vocal(gender, fruit_type, voice_profile):
    genero = _genero_canonico(gender)
    fruta = fruit_type or "Fruit"
    perfil = voice_profile or "consistent expressive character voice"
    if "Female" in genero:
        base = "Feminine / clear adult character voice"
    elif "Male" in genero:
        base = "Masculine / clear adult character voice"
    else:
        base = "Neutral / clear adult character voice"
    return f"{base} / {fruta} / {perfil}"


def criar_voice_registry(elenco, idioma):
    lock = lock_idioma_texto(idioma)
    voices = {}

    for personagem in elenco.get("cast", []):
        nome = personagem.get("name", "CHARACTER")
        speaker_key = chave_speaker(nome)
        voice_model_id = f"{speaker_key}_VOICE_LOCK"
        gender = personagem.get("gender", "")
        fruit_type = personagem.get("fruit_type", "")
        voice_profile = personagem.get("voice_profile", "")
        age = personagem.get("age", "")

        identity_lock = (
            f"Voice Model: {voice_model_id}. "
            f"Speaker: {nome}. Age: {age}. Gender: {gender}. "
            f"Fruit Type: {fruit_type}. Tone: {voice_profile}. {lock}."
        )
        override = {
            "VOICE_GENDER": _genero_canonico(gender),
            "VOCAL_WEIGHT": _peso_vocal(gender, fruit_type, voice_profile),
            "TONE_PROFILE": (
                f"{nome}: {voice_profile}. Keep the same pitch, rhythm, accent, "
                "emotional texture, and vocal weight in every scene."
            ),
            "FORCE_SYNTHESIS": _sintese_forcada(gender, idioma),
        }

        voices[nome] = {
            "SPEAKER": nome,
            "SPEAKER_KEY": speaker_key,
            "VOICE_MODEL_ID": voice_model_id,
            "VOICE_IDENTITY_LOCK": identity_lock,
            "VOICE_OVERRIDE_METADATA": override,
        }

    return {
        "version": "voice-registry-v1",
        "language_lock": lock,
        "rule": (
            "Every DIALOGUE_LINES item must use the exact SPEAKER and "
            "VOICE_IDENTITY_LOCK from this registry. Never invent per-scene voices."
        ),
        "voices": voices,
    }


def obter_voice_registry(elenco, idioma):
    registry = elenco.get("voice_registry")
    if isinstance(registry, dict) and registry.get("voices"):
        return registry
    return criar_voice_registry(elenco, idioma)


def aplicar_voice_registry_elenco(elenco, idioma):
    registry = criar_voice_registry(elenco, idioma)
    elenco["voice_registry"] = registry

    for personagem in elenco.get("cast", []):
        nome = personagem.get("name")
        voz = registry["voices"].get(nome)
        if not voz:
            continue
        personagem["voice_id"] = voz["VOICE_MODEL_ID"]
        personagem["voice_identity_lock"] = voz["VOICE_IDENTITY_LOCK"]
        personagem["voice_override_metadata"] = voz["VOICE_OVERRIDE_METADATA"]

    return elenco


def _mapa_por_chave(registry):
    return {
        chave_speaker(nome): voz
        for nome, voz in registry.get("voices", {}).items()
    }


def encontrar_voz(registry, speaker):
    mapa = _mapa_por_chave(registry)
    return mapa.get(chave_speaker(speaker))


def aplicar_voice_registry_cena(cena, elenco, idioma):
    registry = obter_voice_registry(elenco, idioma)
    overrides = {}

    for linha in cena.get("DIALOGUE_LINES", []):
        if "TEXT" not in linha:
            continue
        voz = encontrar_voz(registry, linha.get("SPEAKER"))
        if not voz:
            continue
        linha["SPEAKER"] = voz["SPEAKER"]
        linha["VOICE_IDENTITY_LOCK"] = voz["VOICE_IDENTITY_LOCK"]
        overrides[voz["SPEAKER"]] = voz["VOICE_OVERRIDE_METADATA"]

    if overrides:
        cena["VOICE_OVERRIDE_METADATA"] = overrides

    return cena


def aplicar_voice_registry_roteiro(roteiro, elenco, idioma):
    roteiro["voice_registry"] = obter_voice_registry(elenco, idioma)
    for cena in roteiro.get("scenes", {}).values():
        aplicar_voice_registry_cena(cena, elenco, idioma)
    return roteiro


def voice_registry_prompt(elenco, idioma):
    registry = obter_voice_registry(elenco, idioma)
    linhas = [
        "VOICE REGISTRY - CANONICO E OBRIGATORIO:",
        f"Language lock: {registry['language_lock']}",
        "Use exatamente estes SPEAKER e VOICE_IDENTITY_LOCK. Nao crie vozes novas.",
    ]
    for nome, voz in registry.get("voices", {}).items():
        linhas.append(f"- SPEAKER: {nome}")
        linhas.append(f"  VOICE_IDENTITY_LOCK: {voz['VOICE_IDENTITY_LOCK']}")
        linhas.append(f"  METADATA: {voz['VOICE_OVERRIDE_METADATA']}")
    return "\n".join(linhas)


def problemas_voice_registry(elenco, roteiro):
    problemas = []
    registry = elenco.get("voice_registry", {})
    if not registry.get("voices"):
        return ["Elenco sem voice_registry canonico."]

    vozes_por_chave = _mapa_por_chave(registry)
    cenas = roteiro.get("scenes", {})

    for chave, cena in cenas.items():
        num = cena.get("SCENE_NUMBER", chave)
        metadata = cena.get("VOICE_OVERRIDE_METADATA", {})
        for linha in cena.get("DIALOGUE_LINES", []):
            if "TEXT" not in linha:
                continue
            speaker = linha.get("SPEAKER", "")
            voz = vozes_por_chave.get(chave_speaker(speaker))
            if not voz:
                problemas.append(f"Cena {num}: SPEAKER fora do elenco/registry: {speaker}")
                continue
            if linha.get("SPEAKER") != voz["SPEAKER"]:
                problemas.append(f"Cena {num}: SPEAKER deveria ser {voz['SPEAKER']}, veio {speaker}")
            if linha.get("VOICE_IDENTITY_LOCK") != voz["VOICE_IDENTITY_LOCK"]:
                problemas.append(f"Cena {num} ({speaker}): VOICE_IDENTITY_LOCK diferente do registry")
            if metadata.get(voz["SPEAKER"]) != voz["VOICE_OVERRIDE_METADATA"]:
                problemas.append(f"Cena {num} ({speaker}): VOICE_OVERRIDE_METADATA diferente do registry")

    return problemas

File: src/voz.php
<?php
declare(strict_types=1);

require_once __DIR__ . '/dialogo.php';

function chave_speaker(string $nome): string
{
    $chave = preg_replace('/[^A-Z0-9]+/', '_', strtoupper($nome));
    return trim((string) $chave, '_');
}

function _genero_canonico(string $gender): string
{
    $texto = strtolower($gender);
    if (preg_match('/female|femin|mulher|girl|woman/', $texto)) {
        return 'Strictly Female';
    }
    if (preg_match('/male|masc|homem|boy|man/', $texto)) {
        return 'Strictly Male';
    }
    return 'Androgynous / Character Voice';
}

function _sintese_forcada(string $gender, string $idioma): string
{
    $genero = _genero_canonico($gender);
    $codigo = strtoupper(normalizar_idioma($idioma));
    if (str_contains($genero, 'Female')) {
        return "Female_Adult_Soprano_{$codigo}";
    }
    if (str_contains($genero, 'Male')) {
        return "Male_Adult_Baritone_{$codigo}";
    }
    return "Adult_Neutral_Character_{$codigo}";
}

function _peso_vocal(string $gender, string $character_type, string $voice_profile): string
{
    $genero = _genero_canonico($gender);
    $tipo = $character_type !== '' ? $character_type : 'Character';
    $perfil = $voice_profile !== '' ? $voice_profile : 'consistent expressive character voice';
    if (str_contains($genero, 'Female')) {
        $base = 'Feminine / clear adult character voice';
    } elseif (str_contains($genero, 'Male')) {
        $base = 'Masculine / clear adult character voice';
    } else {
        $base = 'Neutral / clear adult character voice';
    }
    return "{$base} / {$tipo} / {$perfil}";
}

function criar_voice_registry(array $elenco, string $idioma): array
{
    $lock = lock_idioma_texto($idioma);
    $voices = [];

    foreach ($elenco['cast'] ?? [] as $personagem) {
        $nome = $personagem['name'] ?? 'CHARACTER';
        $speaker_key = chave_speaker($nome);
        $voice_model_id = "{$speaker_key}_VOICE_LOCK";
        $gender = $personagem['gender'] ?? '';
        $character_type = $personagem['character_type'] ?? ($personagem['fruit_type'] ?? '');
        $voice_profile = $personagem['voice_profile'] ?? '';
        $age = $personagem['age'] ?? '';

        $identity_lock = sprintf(
            'Voice Model: %s. Speaker: %s. Age: %s. Gender: %s. Character Type: %s. Tone: %s. %s.',
            $voice_model_id,
            $nome,
            $age,
            $gender,
            $character_type,
            $voice_profile,
            $lock
        );
        $override = [
            'VOICE_GENDER' => _genero_canonico($gender),
            'VOCAL_WEIGHT' => _peso_vocal($gender, (string) $character_type, (string) $voice_profile),
            'TONE_PROFILE' => (
                "{$nome}: {$voice_profile}. Keep the same pitch, rhythm, accent, "
                . 'emotional texture, and vocal weight in every scene.'
            ),
            'FORCE_SYNTHESIS' => _sintese_forcada($gender, $idioma),
        ];

        $voices[$nome] = [
            'SPEAKER' => $nome,
            'SPEAKER_KEY' => $speaker_key,
            'VOICE_MODEL_ID' => $voice_model_id,
            'VOICE_IDENTITY_LOCK' => $identity_lock,
            'VOICE_OVERRIDE_METADATA' => $override,
        ];
    }

    return [
        'version' => 'voice-registry-v1',
        'language_lock' => $lock,
        'rule' => (
            'Every DIALOGUE_LINES item must use the exact SPEAKER and '
            . 'VOICE_IDENTITY_LOCK from this registry. Never invent per-scene voices.'
        ),
        'multi_speaker_rule' => (
            'When a scene has multiple speakers, synthesize one line at a time using the '
            . 'active speaker only. Never blend voices, never reuse another speaker\'s timbre, '
            . 'and never let visual position or camera focus change the assigned voice.'
        ),
        'voices' => $voices,
    ];
}

function obter_voice_registry(array $elenco, string $idioma): array
{
    $registry = $elenco['voice_registry'] ?? null;
    if (is_array($registry) && !empty($registry['voices'])) {
        return $registry;
    }
    return criar_voice_registry($elenco, $idioma);
}

function aplicar_voice_registry_elenco(array $elenco, string $idioma): array
{
    $registry = criar_voice_registry($elenco, $idioma);
    $elenco['voice_registry'] = $registry;

    foreach ($elenco['cast'] ?? [] as $indice => $personagem) {
        $nome = $personagem['name'] ?? null;
        if ($nome === null) {
            continue;
        }
        $voz = $registry['voices'][$nome] ?? null;
        if ($voz === null) {
            continue;
        }
        $elenco['cast'][$indice]['voice_id'] = $voz['VOICE_MODEL_ID'];
        $elenco['cast'][$indice]['voice_identity_lock'] = $voz['VOICE_IDENTITY_LOCK'];
        $elenco['cast'][$indice]['voice_override_metadata'] = $voz['VOICE_OVERRIDE_METADATA'];
    }

    return $elenco;
}

function _mapa_por_chave(array $registry): array
{
    $mapa = [];
    foreach ($registry['voices'] ?? [] as $nome => $voz) {
        $mapa[chave_speaker($nome)] = $voz;
    }
    return $mapa;
}

function encontrar_voz(array $registry, string $speaker): ?array
{
    $mapa = _mapa_por_chave($registry);
    return $mapa[chave_speaker($speaker)] ?? null;
}

function aplicar_voice_registry_cena(array $cena, array $elenco, string $idioma): array
{
    $registry = obter_voice_registry($elenco, $idioma);
    $overrides = [];

    foreach ($cena['DIALOGUE_LINES'] ?? [] as $indice => $linha) {
        if (!is_array($linha) || !array_key_exists('TEXT', $linha)) {
            continue;
        }
        $voz = encontrar_voz($registry, (string) ($linha['SPEAKER'] ?? ''));
        if ($voz === null) {
            continue;
        }
        $cena['DIALOGUE_LINES'][$indice]['SPEAKER'] = $voz['SPEAKER'];
        $cena['DIALOGUE_LINES'][$indice]['VOICE_IDENTITY_LOCK'] = $voz['VOICE_IDENTITY_LOCK'];
        $overrides[$voz['SPEAKER']] = $voz['VOICE_OVERRIDE_METADATA'];
    }

    if ($overrides !== []) {
        $cena['VOICE_OVERRIDE_METADATA'] = $overrides;
    }

    return $cena;
}

function aplicar_voice_registry_roteiro(array $roteiro, array $elenco, string $idioma): array
{
    $roteiro['voice_registry'] = obter_voice_registry($elenco, $idioma);
    foreach ($roteiro['scenes'] ?? [] as $chave => $cena) {
        $roteiro['scenes'][$chave] = aplicar_voice_registry_cena($cena, $elenco, $idioma);
    }
    return $roteiro;
}

function voice_registry_prompt(array $elenco, string $idioma): string
{
    $registry = obter_voice_registry($elenco, $idioma);
    $linhas = [
        'VOICE REGISTRY - CANONICO E OBRIGATORIO:',
        "Language lock: {$registry['language_lock']}",
        'Use exatamente estes SPEAKER e VOICE_IDENTITY_LOCK. Nao crie vozes novas.',
        'Multi-speaker rule: ' . ($registry['multi_speaker_rule'] ?? ''),
    ];
    foreach ($registry['voices'] ?? [] as $nome => $voz) {
        $linhas[] = "- SPEAKER: {$nome}";
        $linhas[] = "  VOICE_IDENTITY_LOCK: {$voz['VOICE_IDENTITY_LOCK']}";
        $linhas[] = '  METADATA: ' . json_encode($voz['VOICE_OVERRIDE_METADATA'], JSON_UNESCAPED_UNICODE);
    }
    return implode("\n", $linhas);
}

function problemas_voice_registry(array $elenco, array $roteiro): array
{
    $problemas = [];
    $registry = $elenco['voice_registry'] ?? [];
    if (empty($registry['voices'])) {
        return ['Elenco sem voice_registry canonico.'];
    }

    $vozes_por_chave = _mapa_por_chave($registry);
    $cenas = $roteiro['scenes'] ?? [];

    foreach ($cenas as $chave => $cena) {
        $num = $cena['SCENE_NUMBER'] ?? $chave;
        $metadata = $cena['VOICE_OVERRIDE_METADATA'] ?? [];
        foreach ($cena['DIALOGUE_LINES'] ?? [] as $linha) {
            if (!is_array($linha) || !array_key_exists('TEXT', $linha)) {
                continue;
            }
            $speaker = $linha['SPEAKER'] ?? '';
            $voz = $vozes_por_chave[chave_speaker($speaker)] ?? null;
            if ($voz === null) {
                $problemas[] = "Cena {$num}: SPEAKER fora do elenco/registry: {$speaker}";
                continue;
            }
            if (($linha['SPEAKER'] ?? null) !== $voz['SPEAKER']) {
                $problemas[] = "Cena {$num}: SPEAKER deveria ser {$voz['SPEAKER']}, veio {$speaker}";
            }
            if (($linha['VOICE_IDENTITY_LOCK'] ?? null) !== $voz['VOICE_IDENTITY_LOCK']) {
                $problemas[] = "Cena {$num} ({$speaker}): VOICE_IDENTITY_LOCK diferente do registry";
            }
            if (($metadata[$voz['SPEAKER']] ?? null) !== $voz['VOICE_OVERRIDE_METADATA']) {
                $problemas[] = "Cena {$num} ({$speaker}): VOICE_OVERRIDE_METADATA diferente do registry";
            }
        }
    }

    return $problemas;
}